### ISSUE TITLE
What happened to physical machines?

### ISSUE THREAD

**Issue by yepyepyep4711** [hashicorp/terraform#23513]
The doc says "physical machines, [...], network switches", but I can't find a provider or documentation pertaining to provisioning an on-premise network of servers, routers and switches. Is this functionality available or not?

(I know https://github.com/hashicorp/terraform/issues/50 exists, but it's been closed over five years ago)

**Comment by pkolyvas** [comment]
Hi @yepyepyep4711 - A good catch. Terraform's goals are unchanged, and many continue to be aspirational. No community providers have appeared to directly answer this specific need. It's unlikely to be something we focus on in the near term internally, but we'd certainly welcome community contributions in this regard.

**Comment by danieldreier** [comment]
I think the correct next step here would be to improve our documentation around what bare metal capabilities exist using terraform. We'll put this in our backlog around documentation.

I haven't personally done bare metal provisioning using terraform, but it includes a [cobbler provisioner]( https://www.terraform.io/docs/providers/cobbler/index.html exists to), which allows you to PXE boot machines. CoreOS [Matchbox](https://coreos.com/blog/matchbox-with-terraform) has similar capabilities. There are also [network providers](https://www.terraform.io/docs/providers/type/network-index.html) that you can use if one matches the hardware you have available.

Are you in the middle of a project here? If you'd like to do a writeup of a bare-metal compute and network provisioning project using terraform I'd be happy to talk through your experiences and maybe do jointly write up a blog post or similar on your experiences.

**Comment by danieldreier** [comment]
I didn't intend to move this to the cobbler provider by labeling it, re-opening

**Comment by yepyepyep4711** [comment]
To clarify, I bumped into this during a training on Terraform. Being able to provision a physical network sounded really nifty. But after some more reading, maybe if Terraform doesn't provide this by now, maybe it should just say loud and clear "cloud only"?
After all, "provisioning" can't truly be done from the command line, you still need to actually plan for, buy, install and connect the hardware, so maybe it's configuration I'm talking about, and that is covered by Ansible unless I'm mistaken.

**Comment by danieldreier** [comment]
@yepyepyep4711 terraform absolutely can provision physical hardware. For example, it's common to use tooling like cobbler to dynamically reprovision machines, the same way you'd provision or de-provision EC2 instances. Terraform has a cobbler provider and other providers for managing physical network gear. It won't walk to the store and and put your hardware in a rack, but the capabilities are absolutely there to re-provision. I definitely agree that some documentation improvements are needed, and I appreciate the feedback that it's not clear how terraform fits into on-prem provisioning. I'm going to close this out because the github issue tracker is primarily here for reporting defects and feature enhancements to terraform core. If you go further down this path and end up setting up a terraform based hardware provisioning system I'd be happy work together on a joint writeup so we can better explain the use case with a real story.

**Comment by yepyepyep4711** [comment]
@danieldreier is there a list of those providers somewhere? I went to cobbler's website and entered "cisco" in the search. No hit. The docs I can find are about PXE and downloading ISOs, not what I'm looking for.
I'm really scratching my head at why such information is so hard to find. If I had to decide on physical infrastructure management solutions, these would be out because of this. My technicians need to focus on the actual job, not spend days finding out how the cool new thing some manager decided on using actually works.
